test/algorithms/is_valid_failure.cpp

```diff
@@ -0,0 +1,979 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+// Unit Test
+
+// Copyright (c) 2015, Oracle and/or its affiliates.
+
+// Contributed and/or modified by Menelaos Karavelas, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
+
+#ifndef BOOST_TEST_MODULE
+#define BOOST_TEST_MODULE test_is_valid_failure
+#endif
+
+#include <iostream>
+#include <string>
+
+#include <boost/variant/variant.hpp>
+
+#include <boost/test/included/unit_test.hpp>
+
+#include <boost/geometry/core/cs.hpp>
+#include <boost/geometry/geometries/geometries.hpp>
+#include <boost/geometry/algorithms/is_valid.hpp>
+#include <boost/geometry/algorithms/reverse.hpp>
+#include <boost/geometry/algorithms/validity_failure_type.hpp>
+#include <boost/geometry/io/wkt/wkt.hpp>
+
+#include <boost/geometry/strategies/strategies.hpp>
+
+#include <from_wkt.hpp>
+
+
+namespace bg = ::boost::geometry;
+
+typedef bg::validity_failure_type failure_type;
+
+typedef bg::model::point<double, 2, bg::cs::cartesian> point_type;
+typedef bg::model::segment<point_type>                 segment_type;
+typedef bg::model::box<point_type>                     box_type;
+typedef bg::model::linestring<point_type>              linestring_type;
+typedef bg::model::multi_linestring<linestring_type>   multi_linestring_type;
+typedef bg::model::multi_point<point_type>             multi_point_type;
+
+
+char const* to_string(failure_type failure)
+{
+    switch (failure)
+    {
+    case bg::no_failure:
+        return "no_failure";
+    case bg::failure_few_points:
+        return "failure_few_points";
+    case bg::failure_wrong_topological_dimension:
+        return "failure_wrong_topological_dimension";
+    case bg::failure_spikes:
+        return "failure_spikes";
+    case bg::failure_duplicate_points:
+        return "failure_duplicate_points";
+    case bg::failure_not_closed:
+        return "failure_not_closed";
+    case bg::failure_self_intersections:
+        return "failure_self_intersections";
+    case bg::failure_wrong_orientation:
+        return "failure_wrong_orientation";
+    case bg::failure_interior_rings_outside:
+        return "failure_interior_rings_outside";
+    case bg::failure_nested_interior_rings:
+        return "failure_nested_interior_rings";
+    case bg::failure_disconnected_interior:
+        return "failure_disconnected_interior";
+    case bg::failure_intersecting_interiors:
+        return "failure_intersecting_interiors";
+    case bg::failure_wrong_corner_order:
+        return "failure_wrong_corner_order";
+    default:
+        return ""; // to avoid warnings (-Wreturn-type)
+    }    
+}
```

Comment:
Is `to_string()` test function really needed when there is identical `validity_failure_type_message()`? This is probably used to test the message but I think it's not really needed, especially if we consider that the message could be extended to e.g. contain some helper data like point coordinates. The `failure_type` would stay the same, so I think it would be sufficient to just compare the failure types and maybe to check if the failure type is consistent with the message, no need to compare the messages.